clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java

```diff
@@ -195,8 +195,16 @@ public boolean isFull() {
         return logBufferBuilder.isFull();
     }
 
-    public void close() {
+    /**
+     * Writes the final metadata in the batch and updates transaction state with the new sequence number for the
+     * TopicPartition
+     * @param transactionState the global transaction state which keeps track of sequence numbers for each TopicPartition.
+     */
+    public void close(TransactionState transactionState) {
```

Comment:
Yes. This API is awkward. I did it because it involved the fewest changes.

If `close` is passed with a null `transactionState` the count is not incremented, so we can handle discards correctly. 

I will definitely look into moving the logic to `RecordAccumulator`. But for now, I think there is more value in getting the end to end basically working, and then iterating from there.